ci: Add SBOM attestations and bump please release

askpt opened the pull request:
<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

This pull request introduces a streamlined process for generating and attesting SBOMs (Software Bill of Materials) for .NET projects, along with updates to permissions and workflow configurations. The changes enhance automation, improve security, and simplify the SBOM generation workflow.

### SBOM Generation and Attestation Enhancements:
* Added a new composite GitHub Action (`sbom-generator`) to generate SBOMs using CycloneDX, upload them to a release, and create attestations. This action supports configurable inputs like `github-token`, `project-name`, and `release-tag` (`.github/actions/sbom-generator/action.yml`).
* Replaced the previous standalone SBOM generation steps in the release workflow with calls to the new `sbom-generator` action for multiple projects (e.g., `OpenFeature`, `OpenFeature.Hosting`, `OpenFeature.DependencyInjection`) (`.github/workflows/release.yml`).

### Workflow and Permissions Updates:
* Updated the `permissions` section in the release workflow to include `id-token: write` and `attestations: write`, enabling secure SBOM attestation and release tagging (`.github/workflows/release.yml`).
* Added installation of the CycloneDX.NET tool (`dotnet tool install`) to the release workflow for SBOM generation (`.github/workflows/release.yml`).

### Configuration Improvements:
* Added a `signoff` field in the `release-please-config.json` to standardize commit sign-offs for release automation (`release-please-config.json`).

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

Fixes #465 

### Notes
<!-- any additional notes for this PR -->

I was triggering the build manually to see if I could generate the SBOMs and attest them. They should be visible here: https://github.com/open-feature/dotnet-sdk/actions/runs/15072156929

askpt committed:
chore: remove sbom job and update attest package step in release workflow

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
feat: enhance release workflow with CycloneDX integration and SBOM generation

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
feat: streamline SBOM generation and upload process for OpenFeature projects

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
fix: update release workflow permissions and improve release-please configuration

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
fix: specify version for CycloneDX installation in release workflow

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
feat: implement sbom-generator action for streamlined SBOM generation and attestation

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action digest to ff0a06e (#473)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
|
[github/codeql-action](https://redirect.github.com/github/codeql-action)
| action | digest | `60168ef` -> `ff0a06e` |

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Enabled.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/open-feature/dotnet-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MC4xMS4xOCIsInVwZGF0ZWRJblZlciI6IjQwLjExLjE4IiwidGFyZ2V0QnJhbmNoIjoibWFpbiIsImxhYmVscyI6WyJyZW5vdmF0ZSJdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update spec digest to edf0deb (#474)

This PR contains the following updates:

| Package | Update | Change |
|---|---|---|
| spec | digest | `d27e000` -> `edf0deb` |

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Enabled.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/open-feature/dotnet-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MC4xMS4xOCIsInVwZGF0ZWRJblZlciI6IjQwLjExLjE4IiwidGFyZ2V0QnJhbmNoIjoibWFpbiIsImxhYmVscyI6WyJyZW5vdmF0ZSJdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update codecov/codecov-action action to v5.4.3 (#475)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
|
[codecov/codecov-action](https://redirect.github.com/codecov/codecov-action)
| action | patch | `v5.4.2` -> `v5.4.3` |

---

### Release Notes

<details>
<summary>codecov/codecov-action (codecov/codecov-action)</summary>

###
[`v5.4.3`](https://redirect.github.com/codecov/codecov-action/blob/HEAD/CHANGELOG.md#v543)

[Compare
Source](https://redirect.github.com/codecov/codecov-action/compare/v5.4.2...v5.4.3)

##### What's Changed

- build(deps): bump github/codeql-action from 3.28.13 to 3.28.17 by
[@&#8203;app/dependabot](https://redirect.github.com/app/dependabot) in
[https://github.com/codecov/codecov-action/pull/1822](https://redirect.github.com/codecov/codecov-action/pull/1822)
- fix: OIDC on forks by
[@&#8203;joseph-sentry](https://redirect.github.com/joseph-sentry) in
[https://github.com/codecov/codecov-action/pull/1823](https://redirect.github.com/codecov/codecov-action/pull/1823)

**Full Changelog**:
https://github.com/codecov/codecov-action/compare/v5.4.2..v5.4.3

</details>

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Enabled.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/open-feature/dotnet-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MC4xMS4xOCIsInVwZGF0ZWRJblZlciI6IjQwLjExLjE4IiwidGFyZ2V0QnJhbmNoIjoibWFpbiIsImxhYmVscyI6WyJyZW5vdmF0ZSJdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
feat: Add OTEL compatible telemetry object builder (#397)

<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR

This pull request introduces a feature for generating telemetry events
to track feature flag evaluations. It adds a new `EvaluationEvent`
class, a builder for creating these events, constants for telemetry
attributes, and metadata definitions. It also includes comprehensive
unit tests to ensure the correctness of the implementation.

### New Feature: Telemetry Event Generation

* **`EvaluationEvent` class**: Added a new class to represent evaluation
events for feature flags, including properties for the event name and
attributes. (`src/OpenFeature/Telemetry/EvaluationEvent.cs`)
* **`EvaluationEventBuilder` class**: Introduced a static builder class
to construct `EvaluationEvent` instances using flag evaluation details
and hook context.
(`src/OpenFeature/Telemetry/EvaluationEventBuilder.cs`)

### Supporting Components

* **Telemetry constants**: Defined a set of constants in
`TelemetryConstants` to standardize attribute keys for
OpenTelemetry-compliant feature flag events.
(`src/OpenFeature/Telemetry/TelemetryConstants.cs`)
* **Flag metadata attributes**: Added `TelemetryFlagMetadata` to define
well-known metadata attributes for telemetry events, such as
`contextId`, `flagSetId`, and `version`.
(`src/OpenFeature/Telemetry/TelemetryFlagMetadata.cs`)

### Unit Tests

* **`EvaluationEventBuilderTests`**: Added unit tests to validate the
behavior of `EvaluationEventBuilder`, including scenarios for handling
errors, missing metadata, and missing attributes.
(`test/OpenFeature.Tests/Telemetry/EvaluationEventBuilderTests.cs`)

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

Fixes #381

---------

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

kylejuliandev committed:
docs: Add AspNetCore sample app (#477)

<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

- Adds simple a AspNetCore web application with a `/welcome` minimal API
that will show one of two messages based the configuration of the
`InMemory` feature provider.

### Related Issues
<!-- add here the GitHub issue that this PR resolves if applicable -->

### Notes
<!-- any additional notes for this PR -->

While the
[README](https://github.com/open-feature/dotnet-sdk/blob/6c44db9237748735a22a162a88f61b2de7f3e9bd/README.md)
for OpenFeature dotnet is pretty comprehensive, I think it can be quite
intimidating to developers who are unfamilar with feature flagging and
OpenFeature. Providing a basic application that can be cloned locally
and run with minimal setup would improve the onboarding and upskill
experience.

Additionally, as a contributor I find myself frequently creating simple
applications to test out changes to OpenFeature. This would save the
effort of stashing local changes and make it easier to more quickly test
or prototype changes.

The example could be extended with endpoints that explore the other
types of feature flags and how you might utilise them within an
application.

### Follow-up Tasks
<!-- anything that is related to this PR but not done here should be
noted under this section -->
<!-- if there is a need for a new issue, please link it here -->

Update the
[README](https://github.com/open-feature/dotnet-sdk/blob/6c44db9237748735a22a162a88f61b2de7f3e9bd/README.md)
with a reference to the sample application to help guide developers.

### How to test
<!-- if applicable, add testing instructions under this section -->

---------

Signed-off-by: Kyle Julian <38759683+kylejuliandev@users.noreply.github.com>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt commented on .github/workflows/release.yml:
```diff
@@ -54,6 +53,9 @@ jobs:
           global-json-file: global.json
           source-url: https://nuget.pkg.github.com/open-feature/index.json
 
+      - name: Install CycloneDX.NET
```
Let me try moving into the workflow. 👍

askpt replied:
It worked: https://github.com/open-feature/dotnet-sdk/actions/runs/15181424919/job/42691515127

If the cyclone is already instaled, it will just output: 
```
Run dotnet tool install --global CycloneDX --version 5.2.0
Tool 'cyclonedx' is already installed.
```

askpt committed:
feat: Move CycloneDX.NET installation to sbom-generator action

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

beeme1mr commented on .github/workflows/release.yml:
```diff
@@ -12,20 +12,19 @@ permissions:
 
     steps:
-      - uses: googleapis/release-please-action@db8f2c60ee802b3748b512940dde88eabd7b7e01 # v3
+      - uses: googleapis/release-please-action@a02a34c4d625f9be7cb89156071d8567266a2445 #v4
```
I'm in favor of switching to v4 but we'll need to be more defensive in the `if` conditions to ensure everything continues to work. FYI @toddbaert

askpt replied:
@beeme1mr Do you have any suggestions on where I should add the `if` conditions in this action?

beeme1mr replied:
Something like this should work.

https://github.com/open-feature/java-sdk-contrib/blob/main/.github/workflows/release-please.yml#L29

beeme1mr commented on .github/workflows/release.yml:
```diff
@@ -68,34 +67,26 @@ jobs:
-      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
+      # Process OpenFeature project
+      - name: Generate and Attest SBOM for OpenFeature
+        uses: ./.github/actions/sbom-generator
```
Are you able to call this action with a normal GitHub token?

askpt replied:
Yes, the GH token is passed as an input in the composite action.

askpt committed:
chore: Cleanup .props file (#476)

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

This pull request introduces several changes to improve project
configuration and clean up unused code. The most significant updates
include enabling implicit usings globally, removing redundant `Nullable`
and `ImplicitUsings` properties from specific project files, and
cleaning up unused `using` directives across multiple source files.

### Project configuration improvements:
* Enabled `ImplicitUsings` globally in `build/Common.props` to
streamline code and reduce boilerplate. Removed redundant
`TreatWarningsAsErrors` configuration for the Release build.
[[1]](diffhunk://#diff-3e9cdf5c2ef721be7342fcd07bd9e3567805d13357c97221e15137225f64e063R10)
[[2]](diffhunk://#diff-3e9cdf5c2ef721be7342fcd07bd9e3567805d13357c97221e15137225f64e063L21-L24)
* Added conditional `Using` directives for test projects in
`build/Common.tests.props`, ensuring proper dependencies for E2E tests.

### Code cleanup:
* Removed redundant `Nullable` and `ImplicitUsings` properties from
`OpenFeature.DependencyInjection.csproj` and
`OpenFeature.Hosting.csproj` as these are now handled globally.
[[1]](diffhunk://#diff-55279cca9fb64343216b7896e3aa5ea96521b091ba68f0b06a2e9430b0a35962L5-L6)
[[2]](diffhunk://#diff-1ceba8f8b16fba8e2b45f9f2f2c78e2e92129949e560f4346df2dd99fdbbd5f4L5-L6)
* Cleaned up unused `using` directives across multiple source files,
including `Api.cs`, `AsyncLocalTransactionContextPropagator.cs`, and
various exception classes, to improve readability and maintainability.
[[1]](diffhunk://#diff-dc150fbd3b7be797470374ee7e38c7ab8a31eb9b6b7a52345e05114c2d32a15eL1-L6)
[[2]](diffhunk://#diff-d9ca58e32d696079f875c51837dfc6ded087b06eb3aef3513f5ea15ebc22c700L1)
[[3]](diffhunk://#diff-a7a382f3d0da52015d1b9e03802692a86c61e7b57580747f31fae37d8dcc5cd6L1)
[[4]](diffhunk://#diff-b9fdde9b61e62d7c474069ea5fc2a43d123ab86d93cb9a6d0673254a64536722L1)
[[5]](diffhunk://#diff-bb4aadb03ea44e3b6b74b83f12b2b1a258e13836bcf815f996addcd5537a478fL1)
[[6]](diffhunk://#diff-89efe60a8b640cc303a38e5b01bc27ad40599e364ff2654a57b7e42138056cdaL1)
[[7]](diffhunk://#diff-7a8cbfba7673f1b69d3d927f60eb4ab0aa548403a118fad9eb43b9a22875dc02L1)
[[8]](diffhunk://#diff-48ae8447bc31a75ecf4bba768a7b68244fbbc9dd5480db9114d9c74fb10fe2efL1)
[[9]](diffhunk://#diff-9a3b5bf7bf3351a6161fc6dd75830bfbaab7aca730cf3f0ae6cd120a76b2f1b1L1)
[[10]](diffhunk://#diff-c4388b2e7252e2e3ac0967dbfdd4647a924cdfc54da229667a0db3613b243a7eL1)
[[11]](diffhunk://#diff-44c88ae43caf99ee733ec911fa85454a96c57d07fc57d2fadd44e12cd7d31cd4L1)
[[12]](diffhunk://#diff-f563baadcf750bb66efaea76d7ec4783320e6efdc45c468effb531c948a2292cL1-L4)
[[13]](diffhunk://#diff-f94bf65f426e13a14f798a8db21a5c9dd3306f5941bde1aba79af3e41421bfc0L1-L3)
[[14]](diffhunk://#diff-96ebc8fc507d0a19d55b9a5cb57b72a0e8058e09f31ee7d0b39e99b00c5029d8L2-L4)
[[15]](diffhunk://#diff-d69e6a4b3c0fb22dcb05a1104cd353a597bc8f54bb29c458ed41d394d8f1c12aL1-L4)
[[16]](diffhunk://#diff-ac87556ad78da36624700cf5bea44dc7cf279c59bb486203b6af9689427a8c9cL1)

---------

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

Commit:
fix: Correct conditional check for release creation in workflow

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.comsigningkey = F1694EC5F3F13898>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

Commit:
fix: Adjust permissions for release jobs in workflow

Signed-off-by: André Silva <2493377+askpt@users.noreply.github.comsigningkey = F1694EC5F3F13898>
Signed-off-by: André Silva <2493377+askpt@users.noreply.github.com>

askpt committed:
Merge branch 'main' into askpt/465-add-sbom-attestations
